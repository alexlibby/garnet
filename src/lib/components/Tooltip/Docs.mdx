import { Canvas, Meta, Story } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as TooltipStories from "./Tooltip.stories";

<Meta of={TooltipStories} />

# Tooltip

Tooltip is the primary component. It has two possible states.

- [Default](#default)
- [Custom HTML](#customhtml)

<a id="default" />

## Default

This is the default version of the Tooltip component.

<Canvas of={TooltipStories.Default} />

<a id="customhtml" />

## Custom HTML

This variant allows us to display HTML in the Tooltip component.

<Canvas of={TooltipStories.CustomHTML} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={TooltipStories} />
