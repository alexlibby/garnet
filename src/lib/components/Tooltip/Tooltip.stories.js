import Tooltip from "../Tooltip/Tooltip.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Notification Components/ToolTip",
  component: Tooltip,
  argTypes: {
    tip: '<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>',
    showHTML: false,
    timeout: "400",
    label: "more info",
  },
  parameters: {
    docs: {
      story: {
        height: "120px",
      },
    },
  },
};

export const Default = {
  args: {
    tip: '<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>',
    showHTML: false,
    timeout: "400",
    label: "more info",
  },
};

Default.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const CustomHTML = {
  args: {
    tip: '<p>This is an informational tooltip with custom markup - to learn more <a href="/tutorial">click here</a></p>',
    showHTML: true,
    timeout: "300",
    label: "more info",
  },
};

CustomHTML.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};
