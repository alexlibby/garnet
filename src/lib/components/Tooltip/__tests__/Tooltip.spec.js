import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent, userEvent } from "@testing-library/svelte";
import Tooltip from "../../Tooltip/Tooltip.svelte";

const TooltipHTML = `<div class="tooltip s-UAv5UqhRPtrl"><div class="s-UAv5UqhRPtrl"><button aria-describedby="tooltip" type="button" class="trigger s-UAv5UqhRPtrl" data-testid="tooltipButton">?</button> <div aria-hidden="true" id="tooltip" role="tooltip" aria-label="more info" class="s-UAv5UqhRPtrl"></div></div></div>`;
expect.extend(AxeMatchers);

describe("Tests for Tooltip", () => {
  const props = {
    tip: '<p>This is an informational tooltip - to learn more <a href="/tutorial">click here</a></p>',
    showHTML: "false",
    timeout: "400",
    label: "more info",
  };

  test("should render properly as plain text", () => {
    const result = render(Tooltip, props);
    expect(() => result).not.toThrow();
  });

  test("should not show any markup when showHTML is true", async () => {
    const result = render(Tooltip, { ...props, showHTML: "true" });
    expect(() => result).not.toThrow();
  });

  test("should appear when hovering over target", async () => {
    render(Tooltip, props);
    const tooltipButton = screen.getByTestId("tooltipButton");

    await fireEvent.click(tooltipButton);
    expect(screen.getByRole("tooltip")).toBeInTheDocument();
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Tooltip({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => TooltipHTML;
    const result = render();

    expect(await axe(result)).toHaveNoViolations();
  });
});
