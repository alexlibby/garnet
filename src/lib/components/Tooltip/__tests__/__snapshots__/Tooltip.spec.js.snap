// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Tooltip > get a snapshot of component 1`] = `"<div class="garnet-tooltip svelte-130t8to"><div><button aria-describedby="tooltip" type="button" class="trigger svelte-130t8to" data-testid="tooltipButton">?</button> <div aria-hidden="true" id="tooltip" role="tooltip" aria-label="more info" class="svelte-130t8to"></div></div></div>"`;
